### DISCUSSION TITLE
Link browser plugin activity to a default branch?

### DISCUSSION THREAD

**Discussion by ghost** [muety/wakapi#532] [Ideas]
Not sure if this makes sense, but would it be feasible to link the browser plugin activity to activity from the IDE, which contains the current branch? I'm doing a lot of frontend work where I would like to track time spend on http://localhost:3000 
Having the right branch would make it possible to include time spend here in a more detailed report about worked hours.

**Comment by muety** [comment]
Hi @jvanveen, thanks for this thought sparker! The WakaTime browser plugin already includes functionality similar to what you're looking for. There is the `<<LAST_PROJECT>>` placeholder, which resolves to whatever your last project in an editor was. See https://github.com/wakatime/browser-wakatime/issues/28 for a detailed discussion on that.

However, Wakapi [intentionally ignores this](https://github.com/muety/wakapi/blob/6946ad5e3c833bab0baf9d06f4defc888e561991/models/heartbeat.go#L44) at the moment and always falls back to _unknown_ as a project for all browsing activity.

The reason is that it's currently either "all" or "nothing". Either all your browsing activity is attributed to the last coding project (which I think undesirable, see https://github.com/wakatime/browser-wakatime/issues/225, https://github.com/wakatime/browser-wakatime/issues/56) or all of it is _unknown_. A third option is to specify a mapping through using the `@@` syntax in the whitelist , see https://github.com/wakatime/browser-wakatime/issues/76#issuecomment-1497150061. But then you'll only be able to track whitelisted sites, which is probably also not very desirable.

Best option would be to have functionality similar to what I proposed [here](https://github.com/wakatime/browser-wakatime/issues/28#issuecomment-1748443078) in the plugin. Before that, I'm reluctant to implement support for `<<LAST_PROJECT>>` in Wakapi for the above reasons.

**Reply by ghost** [reply]
Hi @muety 

Thank you for your detailed explanation! A mapping of whitelisted sites like being proposed in the linked threads would be great. However, the @@project syntax is already supported? (the browser plugin screenshots show similar formatted whitelisted sites)

> But then you'll only be able to track whitelisted sites, which is probably also not very desirable.

Isn't this the main usecase? Keeping track of whitelisted sites, and linking them to a project? I'm probably overlooking something, but having a whitelisted site like http://localhost:3000@@myproject1 would register all time spent on that url to "myproject" right? That's probably enough to register regular frontend work with I suppose. It would be easy enough to start devservers for different projects on different ports; like http://localhost:3001@@myproject2

**Reply by muety** [reply]
> However, the @@project syntax is already supported?

Yes, should be working. Let me know if not. 

> Isn't this the main usecase? Keeping track of whitelisted sites, and linking them to a project? 

Yes, when using the whitelist, you'll be able to **only** track whitelisted sites, nothing else. So if you set up rules for `http://localhost:3000`, browsing to any other website on the internet won't be reflected in your stats. At least in my understanding.

**Reply by ghost** [reply]
Sorry, I'm confusing @@project with the current branch. Tracking time in "myproject" through http://localhost:3000@@myproject seems to work fine, but the time is added to the "unknown" branch. Would something like a <<LAST_BRANCH>> be an option to consider?

**Reply by muety** [reply]
Should be doable! That's something that would have to be supported upstream (in the CLI) first, though. Can you file an issue at https://github.com/wakatime/browser-wakatime?

**Reply by ghost** [reply]
Sure! I created one at https://github.com/wakatime/browser-wakatime/issues/232

**Reply by ghost** [reply]
The author already added it to the browser plugin! Seems it's also already released (3.0.20). I guess the plugin now sends <<LAST_BRANCH>> then?

**Comment by ghost** [comment]
![Screenshot_20231019_091113](https://github.com/muety/wakapi/assets/48799939/a4e358b3-81d8-4371-9a6f-bab11095a58a)
Seems to work :)

**Comment by muety** [comment]
Wow, that was quick! I added #538 to implement support for this and will hopefully get to it soon.

**Reply by ghost** [reply]
Thanks for creating a ticket for this. Would be great to be able to use more detailed browser activity using branches.

**Reply by ghost** [reply]
Nice, I see #538 is already done! I'll test the new version
